File: unity/QuestSLAM-ros2/Assets/QuestSLAM/web/ws/schema.cs
using System;
using UnityEngine;

namespace QuestSLAM.web.dataschema
{
    [Serializable]
    public class Vec3
    {
        public float x;
        public float y;
        public float z;
    }

    [Serializable]
    public class Quat
    {
        public float x;
        public float y;
        public float z;
        public float w;
    }

    [Serializable]
    public class Pose
    {
        public Vec3 pos;
        public Quat rot;
    }

    [Serializable]
    public struct Packet<T> where T : struct
    {
        public string msgType;
        public T data;
        
        public Packet(string type, T packetData)
        {
            msgType = type;
            data = packetData;
        }
    }

    [Serializable]
    public class IncommingPacket
    {
        public string msgType;
        public string data;
        
        public T GetData<T>() where T : struct
        {
            return JsonUtility.FromJson<T>(data);
        }
    }
    

    [Serializable]
    public struct TelemetryPacket
    {
        public bool connectionStatus;
        public float batteryPercentage;
        public int headsetID;
        public string rosConnectionIP;
        public float rosTime;
        public float cpu;
        public float mem;
        public float temp;
        public bool isTracking;
        public float trackingspeed; 
        public float fps;
        public Pose pose;
    }

    [Serializable]
    public struct LogPacket
    {
        public string logType;
        public string data;
    }

    [Serializable]
    public class Config
    {
        public int headsetID;
        public string rosConnectionIP;
        public int trackingspeed;
        public bool toggleCamera;
        public bool AutoStart;
        public bool AprilTagTracking;
    }
}

File: questslam-webui/src/components/QuestSLAM/schema.ts
export interface Config {
    headsetID: number
    rosConnectionIP: string
    trackingspeed: number
    toggleCamera: boolean
    AutoStart: boolean
    AprilTagTracking: boolean
    AprilTagFamily: string
}

export interface AppInfo {
    AppVersion: string
    AppName: string
    BuildDate: string
    HorisionOSVersion: string
    UnityVersion: string
    DeviceModel: string
}

export type Telemetry  = {
    connectionStatus: boolean;
    batteryPercentage: number;
    headsetID: number;
    rosConnectionIP: string;
    rosTime: number;
    cpu: number;
    mem: number;
    temp: number;
    isTracking: boolean;
    trackingspeed: number;
    fps: number;
    pose: {
        pos: {x: number; y: number; z: number};
        rot: {x: number; y: number; z: number; w: number};
    };
};

export interface LogEntry {
  message: string
  stackTrace?: string
  level: 'info' | 'warning' | 'error' | 'web'
  timestamp: string
}